lib/ansible/modules/cloud/amazon/lightsail.py

```diff
@@ -221,259 +222,122 @@ def create_instance(module, client, instance_name):
-    if not region:
-        module.fail_json(msg='region must be specified')
-
-    client = None
-    try:
-        client = boto3_conn(module, conn_type='client', resource='lightsail',
-                            region=region, endpoint=ec2_url, **aws_connect_kwargs)
-    except (botocore.exceptions.ClientError, botocore.exceptions.ValidationError) as e:
-        module.fail_json(msg='Failed while connecting to the lightsail service: %s' % e, exception=traceback.format_exc())
-
-    changed = False
-    state = module.params['state']
-    name = module.params['name']
-
-    if state == 'absent':
-        changed, instance_dict = delete_instance(module, client, name)
-    elif state in ('running', 'stopped'):
-        changed, instance_dict = startstop_instance(module, client, name, state)
-    elif state == 'restarted':
-        changed, instance_dict = restart_instance(module, client, name)
-    elif state == 'present':
-        changed, instance_dict = create_instance(module, client, name)
-
-    module.exit_json(changed=changed, instance=camel_dict_to_snake_dict(instance_dict))
+        inst = find_instance_info(module, client, instance_name)
 
+    wait = module.params.get('wait')
+    if wait:
+        desired_states = [state]
+        wait_for_instance_state(module, client, instance_name, desired_states)
+        inst = find_instance_info(module, client, instance_name, fail_if_not_found=True)
 
-def _find_instance_info(client, instance_name):
-    ''' handle exceptions where this function is called '''
-    inst = None
-    try:
-        inst = client.get_instance(instanceName=instance_name)
-    except botocore.exceptions.ClientError as e:
-        raise
-    return inst['instance']
+    module.exit_json(changed=changed, instance=camel_dict_to_snake_dict(inst))
 
 
 def main():
-    argument_spec = ec2_argument_spec()
-    argument_spec.update(dict(
+
+    argument_spec = dict(
         name=dict(type='str', required=True),
-        state=dict(type='str', default='present', choices=['present', 'absent', 'stopped', 'running', 'restarted']),
+        state=dict(type='str', default='present', choices=['present', 'absent', 'stopped', 'running', 'restarted',
+                                                           'rebooted']),
         zone=dict(type='str'),
         blueprint_id=dict(type='str'),
         bundle_id=dict(type='str'),
         key_pair_name=dict(type='str'),
-        user_data=dict(type='str'),
+        user_data=dict(type='str', default=''),
         wait=dict(type='bool', default=True),
         wait_timeout=dict(default=300, type='int'),
-    ))
-
-    module = AnsibleModule(argument_spec=argument_spec)
+    )
 
-    if not HAS_BOTO3:
-        module.fail_json(msg='Python module "boto3" is missing, please install it')
+    module = AnsibleAWSModule(argument_spec=argument_spec,
+                              required_if=[['state', 'present', ('zone', 'blueprint_id', 'bundle_id', 'key_pair_name')]])
```

Comment:
Personal niggle: boto3 doesn't list keyPairName as a 'required' argument.

I have some use cases where I bake the keys into the AMI and don't want to pass an additional key to the instances...

Reply:
I don't think Lightsail allows custom AMIs, at least from what I've seen in the docs and prodding at the console a bit.  There is a question of if we should have a default to use the region default key when state=present instead of requiring a user specification, generally the way lightsail does keys is different from the way ec2 does keys though and I'm still familiarizing myself with the differences.

Reply:
I was going off my knowledge of EC2, I've not worked with Lightsail.

Fair enough if you can't use custom AMIs (today).  At the same time, adding additional 'requirements' on top of boto3's only risks breaking use cases you might not have thought of.

Reply:
What concerns me is the implication (not directly stated in the docs, at least that I can find) that you can't add/modify a Lightsail instance's keypair after creation (short of ssh'ing to the instance with an existing key and editing auth keys by hand).  **If** this is correct, I feel like we should anticipate users having a problem here and make keypair a required option for create actions.  It may not be boto-y, but it's still something we can easily anticipate and make nicer for our users.  If that's not correct and I'm just missing something about Lightsail key management, then maybe keypair management or a lightsail_key module would be good future improvements (but outside the scope of this PR).

Reply:
@tremble - The `create_instances` method takes in a `blueprintId` parameter that is analogous to an EC2 AMI. The method had a `customImageName` parameter but that is now deprecated.

It does not look like it is possible to change the keypair used by an existing instance. Boto3 does not require the `keyPairName` because when it is absent lightsail uses the default key for the region. Currently, the module does not support using that key AFAICT. Updating the module to add the capability of using the default keypair is quite easy. In fact, now that I think about it, we should absolutely do that.

@jillr - I didn't quite get the connection between being able to update an instance's keypair and making the key_pair_name a required option. (required option..lol). Are you saying we don't want users to unknowingly use the default key and then not be able to change it? The private key from the default keypair can always be downloaded at a later time unlike keypairs created explicitly.

I did a bare bones implementation of the lightsail_keypair module that is being used to create a keypair for every run of the lightsail integration testsuite. I would be happy to flesh it out into a fully featured module.

Reply:
You can set key_pair_name to "LightsailDefaultKeyPair" (`"key_pair_name": "LightsailDefaultKeyPair"`) to use the default lightsail keypair for that region.

Reply:
@prasadkatti : @jillr is mostly worried about protecting folks from making stupid mistakes.  If not setting the Key Pair meant it was *never* possible to get into a host, then there's a case to be made for making it 'required' on the Ansible side (Amazon for example gives you a nice big warning on the console when you do this).  If there was a mechanism for injecting a new key somewhere down the line, then this is less of a worry.

Given that AWS will inject a default key when you don't supply one, I think it's reasonable to err towards not putting an additional 'requirement' in front of boto.

Reply:
Yes, I agree. We should not make it a requirement.